<!DOCTYPE html>
<html lang="en-us" dir="ltr">
<head><script src="/livereload.js?mindelay=10&amp;v=2&amp;port=1313&amp;path=livereload" data-no-instant defer></script>
  <meta charset="utf-8">
<meta name="viewport" content="width=device-width">
<title>SciPipe paper published in GigaScience | Living Systems</title>

    <link rel="stylesheet" href="/css/main.css">


      <script src="/js/main.js"></script>


</head>
<body>
  <header>
    <h1>Living Systems</h1>

  <nav>
    <ul>
    <li>
      <a href="/">Home</a>
    </li>
    <li>
      <a aria-current="true" class="ancestor" href="/posts/">Posts</a>
    </li>
    <li>
      <a href="/tags/">Tags</a>
    </li>
    </ul>
  </nav>


  </header>
  <main>
    
  <h1>SciPipe paper published in GigaScience</h1>

  <img src="selection_999_198.png">

  
  
  <time datetime="2019-04-27T14:48:00&#43;02:00">April 27, 2019</time>

  <p>We just wanted to share that the paper on our Go-based workflow library,
SciPipe, was just published in GigaScience:</p>
<p><a href="https://doi.org/10.1093/gigascience/giz044"><img src="/site/assets/files/1094/selection_999_198.png" alt="">{width=&ldquo;815&rdquo;
height=&ldquo;787&rdquo;}</a></p>
<h2 id="abstract">Abstract</h2>
<p>::: {.section}
::: {.section}</p>
<h3 id="background">Background</h3>
<p>The complex nature of biological data has driven the development of
specialized software tools. Scientific workflow management systems
simplify the assembly of such tools into pipelines, assist with job
automation, and aid reproducibility of analyses. Many contemporary
workflow tools are specialized or not designed for highly complex
workflows, such as with nested loops, dynamic scheduling, and
parametrization, which is common in, e.g., machine learning.
:::</p>
<p>::: {.section}</p>
<h3 id="findings">Findings</h3>
<p>SciPipe is a workflow programming library implemented in the programming
language Go, for managing complex and dynamic pipelines in
bioinformatics, cheminformatics, and other fields. SciPipe helps in
particular with workflow constructs common in machine learning, such as
extensive branching, parameter sweeps, and dynamic scheduling and
parametrization of downstream tasks. SciPipe builds on flow-based
programming principles to support agile development of workflows based
on a library of self-contained, reusable components. It supports running
subsets of workflows for improved iterative development and provides a
data-centric audit logging feature that saves a full audit trace for
every output file of a workflow, which can be converted to other formats
such as HTML, TeX, and PDF on demand. The utility of SciPipe is
demonstrated with a machine learning pipeline, a genomics, and a
transcriptomics pipeline.
:::</p>
<p>::: {.section}</p>
<h3 id="conclusions">Conclusions</h3>
<p>SciPipe provides a solution for agile development of complex and dynamic
pipelines, especially in machine learning, through a flexible
application programming interface suitable for scientists used to
programming or scripting.</p>
<p><strong>Read:</strong>
<a href="https://doi.org/10.1093/gigascience/giz044">doi.org/10.1093/gigascience/giz044</a><br>
<strong>Get started:</strong> <a href="http://scipipe.org">scipipe.org</a><br>
<strong>Get the source:</strong>
<a href="https://github.com/scipipe/scipipe">github.com/scipipe/scipipe</a></p>
<p><strong>Samuel</strong>
:::
:::</p>

  


  </main>
  <footer>
    <p>Copyright 2024. All rights reserved.</p>

  </footer>
</body>
</html>
