<!DOCTYPE html>
<html lang="en-us" dir="ltr">
<head>
  <meta charset="utf-8">
<meta name="viewport" content="width=device-width">
<title>NGS Bioinformatics Course Day 3: New Luigi helper tool, &#34;real-world&#34; NGS pipelines | Living Systems</title>

      <link rel="stylesheet" href="/css/main.min.0dd62280a0c008a6369ab81bd9d63c41bf7e4b5e9377c4a60df43868258f1db5.css" integrity="sha256-DdYigKDACKY2mrgb2dY8Qb9&#43;S16Td8SmDfQ4aCWPHbU=" crossorigin="anonymous">


      <script src="/js/main.23cd0c7d837263b9eaeb96ee2d9ccfa2969daa3fa00fa1c1fe8701a9b87251a1.js" integrity="sha256-I80MfYNyY7nq65buLZzPopadqj&#43;gD6HB/ocBqbhyUaE=" crossorigin="anonymous"></script>



</head>
<body>
  <header>
    <h1>Living Systems</h1>

  <nav>
    <ul>
    <li>
      <a href="/">Home</a>
    </li>
    <li>
      <a aria-current="true" class="ancestor" href="/posts/">Posts</a>
    </li>
    <li>
      <a href="/tags/">Tags</a>
    </li>
    </ul>
  </nav>


  </header>
  <main>
    
  <h1>NGS Bioinformatics Course Day 3: New Luigi helper tool, &#34;real-world&#34; NGS pipelines</h1>

  
  
  <time datetime="2015-03-03T20:45:00&#43;01:00">March 3, 2015</time>

  <p><p class="image">
    <img src="ngsintro-coding.jpg" alt=""  />
</p>
</p>
<p>It turned out I didn&rsquo;t have the time and strength to blog every day at
the NGS Bioinformatics Intro course, so here comes a wrap up with some
random notes and tidbits from the last days, including any concluding
remarks!</p>
<p>These days we started working on a more realistic NGS pipeline, on
analysing re-sequencing samples
(<a href="http://uppnex.se/twiki/pub/Courses/NgsIntro1502/Schedule/NGS_course_AJ_20150211.pdf" target="_blank" rel="noopener">slides</a>
,
<a href="http://uppnex.se/twiki/do/view/Courses/NgsIntro1502/ResequencingAnalysis" target="_blank" rel="noopener">tutorial</a>
).</p>
<h2 id="first-some-outcome-from-this-tutorial">First some outcome from this tutorial</h2>
<p>What do I mean with &ldquo;outcome&rdquo;? Well, as I tried to manually copy and
paste the <a href="http://uppnex.se/twiki/do/view/Courses/NgsIntro1502/ResequencingAnalysis.html" target="_blank" rel="noopener">bag of hairy nasty long bash commandline strings in the
tutorial
pages</a>
,
that all depended upon each other, I got so frustrated that I decided to
try to encode them in a workflow language / tool.</p>
<p>I quickly tried out
<a href="https://bitbucket.org/johanneskoester/snakemake" target="_blank" rel="noopener">SnakeMake</a>
, and was in
fact impressed by the quality and robustness of this tool, but I figured
it would take too long to properly grok the make-inspired way of
thinking in this tool, for me to finish the labs in time. Instead I set
out to use <a href="https://github.com/spotify/luigi" target="_blank" rel="noopener">spotify&rsquo;s luigi tool</a>
,
which I&rsquo;m intimately familiar with already.</p>
<p>Only problem now, was that the relatively extreme amounts of boilerplate
that needs to be written for each luigi task (really it&rsquo;s not that bad,
but compared to just running a shell command, it is).</p>
<p>This resulted in that I realised some ideas I&rsquo;ve been pondering on for
a long time, on how a little bit of Flow-based programming inspired
ideas could be implemented in Luigi, and so I created a little (50 LOC)
<a href="https://github.com/samuell/luigis_monkey_wrench" target="_blank" rel="noopener">helper library</a>
 for
doing just this! <a href="https://github.com/samuell/luigis_monkey_wrench" target="_blank" rel="noopener">The lib, called &ldquo;Luigi&rsquo;s monkey wrench&rdquo; is
available on github</a>
.</p>
<p>And, of course I had to encode the tutorial in it too! You can find the
<a href="https://gist.github.com/samuell/6da9a7c1e03912fde62e" target="_blank" rel="noopener">luigi&rsquo;s-monkey-wrench code for the tutorial in this
gist</a>
!</p>
<p>Let&rsquo;s just include a small snippet, to give you a taste for how the
code looks with this helper:</p>
<div class="highlight"><pre tabindex="0" style="background-color:#fff;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span><span style="color:#00f">import</span> luigi
</span></span><span style="display:flex;"><span><span style="color:#00f">from</span> luigis_monkey_wrench <span style="color:#00f">import</span> *
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>REF=<span style="color:#a31515">&#39;human_17_v37.fasta&#39;</span>
</span></span><span style="display:flex;"><span>INDIVIDUALS=[<span style="color:#a31515">&#39;NA06984&#39;</span>,<span style="color:#a31515">&#39;NA07000&#39;</span>]
</span></span><span style="display:flex;"><span>SAMPLES=[<span style="color:#a31515">&#39;1&#39;</span>,<span style="color:#a31515">&#39;2&#39;</span>]
</span></span><span style="display:flex;"><span>BASENAME=<span style="color:#a31515">&#39;.ILLUMINA.low_coverage.17q_&#39;</span>
</span></span><span style="display:flex;"><span>PICARDDIR=<span style="color:#a31515">&#39;/sw/apps/bioinfo/picard/1.69/kalkyl/&#39;</span>
</span></span><span style="display:flex;"><span>KNOWNSITES=(<span style="color:#a31515">&#39;/proj/g2014207/labs/gatk/&#39;</span>
</span></span><span style="display:flex;"><span>            <span style="color:#a31515">&#39;ALL.chr17.phase1_integrated_calls.20101123.snps_indels_svs.genotypes.vcf&#39;</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#00f">class</span> <span style="color:#2b91af">GATKWorkflow</span>(WorkflowTask):
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>    <span style="color:#00f">def</span> requires(self):
</span></span><span style="display:flex;"><span>        <span style="color:#00f">for</span> i <span style="color:#00f">in</span> INDIVIDUALS:
</span></span><span style="display:flex;"><span>            <span style="color:#008000"># Workflow definition</span>
</span></span><span style="display:flex;"><span>            <span style="color:#008000"># ---------------------------------------------------------------------</span>
</span></span><span style="display:flex;"><span>            <span style="color:#008000"># files() will return a pseudo task that just outputs an existing file,</span>
</span></span><span style="display:flex;"><span>            <span style="color:#008000">#         while not running anything.</span>
</span></span><span style="display:flex;"><span>            <span style="color:#008000"># shell() will create a new task with a command that can take inputs</span>
</span></span><span style="display:flex;"><span>            <span style="color:#008000">#         and outputs.</span>
</span></span><span style="display:flex;"><span>            fq1 = file(<span style="color:#a31515">&#39;fastq:</span><span style="color:#a31515">{i}</span><span style="color:#a31515">/</span><span style="color:#a31515">{i}{b}</span><span style="color:#a31515">1.fq&#39;</span>.format(i=i,b=BASENAME))
</span></span><span style="display:flex;"><span>            fq2 = file(<span style="color:#a31515">&#39;fastq:</span><span style="color:#a31515">{i}</span><span style="color:#a31515">/</span><span style="color:#a31515">{i}{b}</span><span style="color:#a31515">2.fq&#39;</span>.format(i=i,b=BASENAME))
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>            <span style="color:#008000"># Step 2 in [1]--------------------------------------------------------</span>
</span></span><span style="display:flex;"><span>            aln1 = shell(<span style="color:#a31515">&#39;bwa aln </span><span style="color:#a31515">{ref}</span><span style="color:#a31515"> &lt;i:fastq&gt; &gt; &lt;o:sai:&lt;i:fastq&gt;.sai&gt;&#39;</span>.format(
</span></span><span style="display:flex;"><span>                ref=REF))
</span></span><span style="display:flex;"><span>            aln1.inports[<span style="color:#a31515">&#39;fastq&#39;</span>] = fq1.outport(<span style="color:#a31515">&#39;fastq&#39;</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>            aln2 = shell(<span style="color:#a31515">&#39;bwa aln </span><span style="color:#a31515">{ref}</span><span style="color:#a31515"> &lt;i:fastq&gt; &gt; &lt;o:sai:&lt;i:fastq&gt;.sai&gt;&#39;</span>.format(
</span></span><span style="display:flex;"><span>                ref=REF))
</span></span><span style="display:flex;"><span>            aln2.inports[<span style="color:#a31515">&#39;fastq&#39;</span>] = fq2.outport(<span style="color:#a31515">&#39;fastq&#39;</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>            <span style="color:#008000"># ... and so the script continues ...</span>
</span></span></code></pre></div><p>For a closer look, have a look at:</p>
<ul>
<li><a href="https://github.com/samuell/luigis_monkey_wrench" target="_blank" rel="noopener">The tool on
github</a>
</li>
<li><a href="https://gist.github.com/samuell/6da9a7c1e03912fde62e" target="_blank" rel="noopener">The full resulting code from this
tutorial</a>
</li>
</ul>
<h2 id="overall-structure-of-tutorial">Overall structure of tutorial</h2>
<p>The overall structure of the task was something like:</p>
<ol>
<li>Setup data</li>
<li>Map reads to a reference (generating .bam file)</li>
<li>Re-calibrate (new .bam file)</li>
<li>Identify (&ldquo;call&rdquo;) variants (resulting in .vcf file, &ldquo;variant
calling file&rdquo;)</li>
</ol>
<h3 id="the-mapping-step">The mapping step</h3>
<p>Regarding the mapping step, it was mentioned about some popular mapping
tools:</p>
<ul>
<li><a href="http://maq.sourceforge.net/" target="_blank" rel="noopener">Maq</a>
</li>
<li><a href="http://bio-bwa.sourceforge.net/" target="_blank" rel="noopener">BWA</a>
</li>
<li><a href="http://bowtie-bio.sourceforge.net/index.shtml" target="_blank" rel="noopener">Bowtie</a>
</li>
<li><a href="https://code.google.com/p/mosaik-aligner/" target="_blank" rel="noopener">Mosaik</a>
</li>
</ul>
<p>&hellip; then, interestingly, it was mentioned a bit of what kind of
computation tricks they use to do the mapping in an efficient way:</p>
<ul>
<li>Hashtable (Maq, Eland, SOAP, Shrimp, Zoom)</li>
<li>Suffix trees: (BWA, &hellip;)</li>
<li>Burrows-Wheeler transofrm, for file compression (Many tools?)</li>
</ul>
<h3 id="the-re-calibration-step">The re-calibration step</h3>
<p>For the re-calibration step, the more detailed steps was:</p>
<ol>
<li>Local re-aliagnment
<ol>
<li>(Loosely defined as: &ldquo;Using all the reads in a section of the
genome, to get a better hint of which variant is probably the
&lsquo;correct&rsquo; one&rdquo;)</li>
</ol>
</li>
<li>Removal of duplicates</li>
<li>Re-calibration (adjust for biases)
<ol>
<li>Reported vs. empirical quality score</li>
<li>(Biaes from which cycle in the machine etc)</li>
</ol>
</li>
</ol>
<p>Typical software in this section was:</p>
<ul>
<li><a href="http://samtools.sourceforge.net/" target="_blank" rel="noopener">Samtools</a>
 - Tools for working
with the Short-read Alignments (SAM) format (and the binary BAM
counterpart)</li>
<li><a href="http://samtools.sourceforge.net/" target="_blank" rel="noopener">Picard</a>
 - Java based tool for
working with SAM/BAM formats</li>
<li><a href="https://www.broadinstitute.org/gatk/" target="_blank" rel="noopener">GATK</a>
 - Genome Analysis
Toolkit</li>
<li><a href="http://www.broadinstitute.org/igv/" target="_blank" rel="noopener">IGV</a>
 - Integrated Genomics
Viewer</li>
</ul>
<h3 id="the-variant-calling-step">The variant calling step</h3>
<p>Some unsorted notes:</p>
<ul>
<li><a href="https://github.com/ekg/freebayes" target="_blank" rel="noopener">FreeBayes</a>
 - Bayesian SNP calling</li>
<li>Bayesian methods instead of a simple pileup.</li>
<li>Population-based calling.</li>
</ul>
<h2 id="discovery-of-structural-variants">Discovery of structural variants</h2>
<p>We had some discussion on discovery of structural variants too. Find
some random notes below:</p>
<ul>
<li>Harder (than SNP calling)</li>
<li>No perfect way to do this</li>
<li>Different methods depending on data and type of variant that you
want to find.</li>
</ul>
<p>Some methods used to do this:</p>
<ul>
<li>Read depth analysis <em>(Differences in coverage depth (how many times
a certain sequence appear to have been sequenced) might indicate
duplicattions etc)</em></li>
<li>Mate pair sequencing <em>(To know the approximate distance between
&ldquo;pair&rdquo; of known sequences, on longer distances than the typical
read length)</em></li>
<li>Long-reads <em>(Goes without saying, that longer reads will identify
large-scale differences easier &hellip; the problem is just to get long
enough reads)</em></li>
<li>De Novo assembly <em>(Really hard, but when it works, it gives a really
clear picture)</em></li>
</ul>
<h2 id="see-abbreviations">See abbreviations</h2>
<ul>
<li>PGM - Personal Genome Machine</li>
<li>WGS - Whole Genome Study (as opposed to Exome study, which skips the
Introns)</li>
</ul>
<h2 id="common-commands">Common commands</h2>
<div class="highlight"><pre tabindex="0" style="background-color:#fff;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-bash" data-lang="bash"><span style="display:flex;"><span>bwa aln -a bwtsw <span style="color:#008000"># Align with bwtsw algorithm</span>
</span></span><span style="display:flex;"><span>samtools faidx
</span></span><span style="display:flex;"><span>bwa sampe        <span style="color:#008000"># Generate alignments in SAM format given paired-end reads</span>
</span></span></code></pre></div><h2 id="some-common-file-types">Some common file types</h2>
<ul>
<li><code>.fastq</code> - FASTA format with quality scores.</li>
<li><code>.sam</code> - Short-read alignment file format</li>
<li><code>.bam</code> - Binary SAM format (for compression)</li>
<li><code>.sai</code> - SAM file index</li>
<li><code>.gtf / .gff / .bed</code> - Annotation file formats</li>
</ul>
<h2 id="see-also">See also</h2>
<ul>
<li><a href="/posts/ngs-intro-course-day-1/">NGS Bioinformatics Intro Course Day 1</a>
</li>
<li><a href="/posts/ngs-bioinformatics-intro-course-day-2/">NGS Bioinformatics Intro Course Day 2</a>
</li>
</ul>

  
  <div>
    <div>Tags:</div>
    <ul>
        <li><a href="/tags/bioinformatics/">Bioinformatics</a></li>
        <li><a href="/tags/genomics/">Genomics</a></li>
    </ul>
  </div>


  </main>
  <footer>
    <p>Copyright &copy; 2024 Living Systems. All rights reserved.</p>

  </footer>
</body>
</html>
